internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,359 @@
+type imageMirroringMode string
+
+const (
+	// imageMirrorModeJournal uses journaling to propagate RBD images between
+	// ceph clusters.
+	imageMirrorModeJournal imageMirroringMode = "journal"
+	// imageMirrorModeSnapshot uses snapshots to propagate RBD images between
+	// ceph clusters.
+	imageMirrorModeSnapshot imageMirroringMode = "snapshot"
+)
+
+// imageMirroingState represents the image mirroring state.
+type imageMirroingState string
+
+const (
+	// If the state is up+replaying, then mirroring is functioning properly.
+	// up means the rbd-mirror daemon is running, and replaying means
+	// this image is the target for replication from another storage cluster.
+	upAndReplaying imageMirroingState = "up+replaying"
+	// If the state is up+stopped means the rbd-mirror daemon is
+	// running and stopped means the image is not a target for replication from
+	// another cluster
+	upAndStopped imageMirroingState = "up+stopped"
+)
+
+// mirroringMode is the key to get the imageMirroringMode from parameters.
+var imageMirroringKey = "mirroringMode"
+
+// forceKey is the key to get the force option from parameters.
+var forceKey = "force"
+
+// getVolumeFromID gets the rbd image details from the volumeID.
+func (cs *ControllerServer) getVolumeFromID(ctx context.Context, volumeID string, secrets map[string]string) (*rbdVolume, error) {
+	// validate the volume ID
+	cr, err := util.NewUserCredentials(secrets)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if volumeID == "" {
+		return nil, status.Error(codes.InvalidArgument, "empty volume ID in request")
+	}
+
+	if acquired := cs.VolumeLocks.TryAcquire(volumeID); !acquired {
+		util.ErrorLog(ctx, util.VolumeOperationAlreadyExistsFmt, volumeID)
+		return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volumeID)
+	}
+
+	var rbdVol = &rbdVolume{}
+	rbdVol, err = genVolFromVolID(ctx, volumeID, cr, secrets)
+	if err != nil {
+		switch {
+		case errors.Is(err, ErrImageNotFound):
+			err = status.Errorf(codes.NotFound, "volume %s not found", volumeID)
+		case errors.Is(err, util.ErrPoolNotFound):
+			err = status.Errorf(codes.NotFound, "pool %s not found for %s", rbdVol.Pool, volumeID)
+		default:
+			err = status.Errorf(codes.Internal, err.Error())
+		}
+	}
+	return rbdVol, err
+}
+
+// getForceOption extracts the force option from the parameters.
+// If not set, the default will be set to false.
+func getForceOption(ctx context.Context, parameters map[string]string) (bool, error) {
+	val, ok := parameters[forceKey]
+	if !ok {
+		util.WarningLog(ctx, "%s is not set in parameters, setting to mirroringMode to default (%v)", forceKey, false)
+		return false, nil
+	}
+	force, err := strconv.ParseBool(val)
+	if err != nil {
+		return false, status.Errorf(codes.Internal, err.Error())
+	}
+	return force, nil
+}
+
+func getMirroringMode(ctx context.Context, parameters map[string]string) (librbd.ImageMirrorMode, error) {
```

Comment:
make sure to document things, specially functions that receive a map as parameter (a map is usually not a nice API)

Reply:
the map is from the request where the user sets the opaque key-value, documented it for now